File: app/src/main/java/net/feheren_fekete/applist/applistpage/viewmodel/ViewModelUtils.java
package net.feheren_fekete.applist.applistpage.viewmodel;

import net.feheren_fekete.applist.applistpage.model.AppData;
import net.feheren_fekete.applist.applistpage.model.ApplistModel;
import net.feheren_fekete.applist.applistpage.model.PageData;
import net.feheren_fekete.applist.applistpage.model.SectionData;
import net.feheren_fekete.applist.applistpage.model.ShortcutData;
import net.feheren_fekete.applist.applistpage.model.StartableData;

import java.util.ArrayList;
import java.util.List;

public class ViewModelUtils {

    public static List<BaseItem> modelToView(ApplistModel applistModel, PageData pageData) {
        List<BaseItem> result = new ArrayList<>();
        for (SectionData sectionData : pageData.getSections()) {
            result.add(new SectionItem(
                    sectionData.getId(),
                    sectionData.getName(),
                    sectionData.isRemovable(),
                    sectionData.isCollapsed()));
            for (StartableData startableData : sectionData.getStartables()) {
                if (startableData instanceof AppData) {
                    AppData appData = (AppData) startableData;
                    result.add(new AppItem(
                            appData.getId(),
                            appData.getPackageName(),
                            appData.getClassName(),
                            appData.getName()));
                } else if (startableData instanceof ShortcutData) {
                    ShortcutData shortcutData = (ShortcutData) startableData;
                    result.add(new ShortcutItem(
                            shortcutData.getId(),
                            shortcutData.getName(),
                            shortcutData.getIntent(),
                            applistModel.getShortcutIconPath(shortcutData)));
                }
            }
        }
        return result;
    }

    public static PageData viewToModel(long pageId, String pageName, List<BaseItem> items) {
        List<SectionData> sectionDatas = new ArrayList<>();
        SectionData sectionData = null;
        List<StartableData> startableDatas = new ArrayList<>();
        for (BaseItem item : items) {
            if (item instanceof SectionItem) {
                SectionItem sectionItem = (SectionItem) item;
                if (sectionData != null) {
                    sectionDatas.add(sectionData);
                }
                // Start a new section
                startableDatas = new ArrayList<>();
                sectionData = new SectionData(
                        sectionItem.getId(),
                        sectionItem.getName(),
                        startableDatas,
                        sectionItem.isRemovable(),
                        sectionItem.isCollapsed());
            } else if (item instanceof AppItem) {
                AppItem appItem = (AppItem) item;
                startableDatas.add(new AppData(
                        appItem.getId(),
                        appItem.getPackageName(),
                        appItem.getClassName(),
                        appItem.getName()));
            } else if (item instanceof ShortcutItem) {
                ShortcutItem shortcutItem = (ShortcutItem) item;
                startableDatas.add(new ShortcutData(
                        shortcutItem.getId(),
                        shortcutItem.getName(),
                        shortcutItem.getIntent()));
            }
        }
        if (sectionData != null) {
            sectionDatas.add(sectionData);
        }
        return new PageData(pageId, pageName, sectionDatas);
    }

}

File: app/src/main/java/net/feheren_fekete/applist/applistpage/viewmodel/ViewModelUtils.kt
package net.feheren_fekete.applist.applistpage.viewmodel

import net.feheren_fekete.applist.applistpage.model.AppData
import net.feheren_fekete.applist.applistpage.model.AppShortcutData
import net.feheren_fekete.applist.applistpage.model.ApplistModel
import net.feheren_fekete.applist.applistpage.model.PageData
import net.feheren_fekete.applist.applistpage.model.SectionData
import net.feheren_fekete.applist.applistpage.model.ShortcutData
import net.feheren_fekete.applist.applistpage.model.StartableData

import java.util.ArrayList
import java.util.Collections

object ViewModelUtils {

    fun modelToView(applistModel: ApplistModel, pageData: PageData): List<BaseItem> {
        val result = ArrayList<BaseItem>()
        for (sectionData in pageData.sections) {
            result.add(SectionItem(
                    sectionData.id,
                    sectionData.name,
                    sectionData.isRemovable,
                    sectionData.isCollapsed))
            for (startableData in sectionData.getStartables()) {
                if (startableData is AppData) {
                    result.add(AppItem(
                            startableData.id,
                            startableData.packageName,
                            startableData.className,
                            startableData.versionCode,
                            startableData.name,
                            startableData.customName,
                            applistModel.getCustomAppIconPath(startableData)))
                } else if (startableData is ShortcutData) {
                    result.add(ShortcutItem(
                            startableData.id,
                            startableData.name,
                            startableData.customName,
                            applistModel.getCustomShortcutIconPath(startableData),
                            startableData.intent,
                            applistModel.getShortcutIconPath(startableData)))
                } else if (startableData is AppShortcutData) {
                    result.add(AppShortcutItem(
                            startableData.id,
                            startableData.name,
                            startableData.customName,
                            applistModel.getCustomShortcutIconPath(startableData),
                            startableData.packageName,
                            startableData.shortcutId,
                            applistModel.getShortcutIconPath(startableData)))
                }
            }
        }
        return result
    }

    fun viewToModel(pageId: Long, pageName: String, items: List<BaseItem>): PageData {
        val sectionDatas = ArrayList<SectionData>()
        var sectionData: SectionData? = null
        var startableDatas: MutableList<StartableData> = ArrayList()
        for (item in items) {
            if (item is SectionItem) {
                // Finish previous section
                if (sectionData != null) {
                    sectionData.setStartables(startableDatas)
                    sectionDatas.add(sectionData)
                }
                // Start a new section
                startableDatas = ArrayList()
                sectionData = SectionData(
                        item.id,
                        item.name,
                        emptyList(),
                        item.isRemovable,
                        item.isCollapsed)
            } else if (item is AppItem) {
                startableDatas.add(AppData(
                        item.id,
                        item.packageName,
                        item.className,
                        item.versionCode,
                        item.name,
                        item.customName))
            } else if (item is ShortcutItem) {
                (item.intent.getPackage()
                        ?: item.intent.component?.packageName)?.let {
                    startableDatas.add(ShortcutData(
                            item.id,
                            it,
                            item.name,
                            item.customName,
                            item.intent))
                }
            } else if (item is AppShortcutItem) {
                startableDatas.add(AppShortcutData(
                        item.id,
                        item.name,
                        item.customName,
                        item.packageName,
                        item.shortcutId))
            }
        }
        if (sectionData != null) {
            sectionData.setStartables(startableDatas)
            sectionDatas.add(sectionData)
        }
        return PageData(pageId, pageName, sectionDatas)
    }

}
